File: onboarder/src/main/java/com/chyrta/onboarder/OnboarderActivity.java
package com.chyrta.onboarder;

import android.animation.ArgbEvaluator;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;

import com.chyrta.onboarder.utils.ColorsArrayBuilder;
import com.chyrta.onboarder.views.CircleIndicatorView;

import java.util.List;

public abstract class OnboarderActivity extends AppCompatActivity implements View.OnClickListener, ViewPager.OnPageChangeListener {

    private Integer[] colors;
    private CircleIndicatorView circleIndicatorView;
    private ViewPager vpOnboarderPager;
    private OnboarderAdapter onboarderAdapter;
    private ImageButton ibNext;
    private Button btnSkip, btnFinish;
    private ArgbEvaluator evaluator;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_onboarder);
        setStatusBackgroundColor();
        hideActionBar();
        circleIndicatorView = (CircleIndicatorView) findViewById(R.id.circle_indicator_view);
        ibNext = (ImageButton) findViewById(R.id.ib_next);
        btnSkip = (Button) findViewById(R.id.btn_skip);
        btnFinish = (Button) findViewById(R.id.btn_finish);
        vpOnboarderPager = (ViewPager) findViewById(R.id.vp_onboarder_pager);
        vpOnboarderPager.addOnPageChangeListener(this);
        ibNext.setOnClickListener(this);
        btnSkip.setOnClickListener(this);
        btnFinish.setOnClickListener(this);
        evaluator = new ArgbEvaluator();
    }

    public void setOnboardPagesReady(List<OnboarderPage> pages) {
        onboarderAdapter = new OnboarderAdapter(pages, getSupportFragmentManager());
        vpOnboarderPager.setAdapter(onboarderAdapter);
        colors = ColorsArrayBuilder.getPageBackgroundColors(this, pages);
        circleIndicatorView.setPageIndicators(pages.size());
    }

    public void setInactiveIndicatorColor(int color) {
        circleIndicatorView.setInactiveIndicatorColor(color);
    }

    public void setActiveIndicatorColor(int color) {
        circleIndicatorView.setActiveIndicatorColor(color);
    }

    public void setStatusBackgroundColor() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().getDecorView().setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                            | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
            getWindow().setStatusBarColor(ContextCompat.getColor(this, R.color.black_transparent));
        }
    }

    @Override
    public void onClick(View v) {
        int i = v.getId();
        if (i == R.id.ib_next) {
            vpOnboarderPager.setCurrentItem(vpOnboarderPager.getCurrentItem() + 1);
        } else if (i == R.id.btn_skip) {
            vpOnboarderPager.setCurrentItem(onboarderAdapter.getCount());
            onSkipButtonPressed();
        } else if (i == R.id.btn_finish) {
            onFinishButtonPressed();
        }
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        if(position < (onboarderAdapter.getCount() - 1) && position < (colors.length - 1)) {
            vpOnboarderPager.setBackgroundColor((Integer) evaluator.evaluate(positionOffset, colors[position], colors[position + 1]));
        } else {
            vpOnboarderPager.setBackgroundColor(colors[colors.length - 1]);
        }
    }

    @Override
    public void onPageSelected(int position) {
        int lastPagePosition = onboarderAdapter.getCount() - 1;
        circleIndicatorView.setCurrentPage(position);
        ibNext.setVisibility(position == lastPagePosition ? View.GONE : View.VISIBLE);
        btnFinish.setVisibility(position == lastPagePosition ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    private void hideActionBar() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }
    }

    abstract public void onSkipButtonPressed();
    abstract public void onFinishButtonPressed();

}

File: onboarder/src/main/kotlin/com/chyrta/onboarder/OnboarderActivity.kt
package com.chyrta.onboarder

import android.content.res.ColorStateList
import android.os.Bundle
import androidx.annotation.ColorInt
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.ColorUtils
import androidx.fragment.app.Fragment
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.chyrta.onboarder.extensions.onLastPage
import com.chyrta.onboarder.extensions.scrollToNextPage
import com.chyrta.onboarder.views.CircleIndicatorView
import com.google.android.material.floatingactionbutton.FloatingActionButton

abstract class OnboarderActivity : AppCompatActivity() {
    private lateinit var circleIndicatorView: CircleIndicatorView
    private lateinit var viewPager: ViewPager2
    private lateinit var fab: FloatingActionButton

    private lateinit var colors: List<Int>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_onboarder)

        circleIndicatorView = findViewById(R.id.circle_indicator_view)
        fab = findViewById(R.id.fab)
        viewPager = findViewById<ViewPager2>(R.id.vp_onboarder_pager)
            .apply {
                registerOnPageChangeCallback(
                    object : ViewPager2.OnPageChangeCallback() {

                        override fun onPageScrolled(
                            position: Int,
                            positionOffset: Float,
                            positionOffsetPixels: Int
                        ) {
                            viewPager.setBackgroundColor(
                                if (!viewPager.onLastPage(position))
                                    ColorUtils.blendARGB(
                                        colors[position],
                                        colors[position + 1],
                                        positionOffset
                                    )
                                else
                                    colors.last()
                            )
                        }

                        override fun onPageSelected(position: Int) {
                            circleIndicatorView.setCurrentPage(position)

                            fab.setImageResource(
                                if (onLastPage(position))
                                    R.drawable.ic_done_24
                                else
                                    R.drawable.ic_arrow_forward_24
                            )
                        }
                    }
                )
            }

        setOnClickListeners()
    }

    private fun setOnClickListeners() {
        fab.setOnClickListener {
            if (!viewPager.onLastPage)
                viewPager.scrollToNextPage()
            else
                onFinishButtonPressed()
        }
    }

    abstract fun onFinishButtonPressed()

    @Suppress("unused")
    fun setPages(pages: List<OnboarderPage>) {
        viewPager.adapter = object : FragmentStateAdapter(this) {
            override fun createFragment(position: Int): Fragment =
                OnboarderFragment.newInstance(pages[position])

            override fun getItemCount(): Int =
                pages.size
        }
        colors = pages.backgroundColors(this)
        circleIndicatorView.setPageIndicators(pages.size)
    }

    @Suppress("unused")
    fun setInactiveIndicatorColor(color: Int) {
        circleIndicatorView.setInactiveIndicatorColor(color)
    }

    @Suppress("unused")
    fun setActiveIndicatorColor(color: Int) {
        circleIndicatorView.setActiveIndicatorColor(color)
    }

    @Suppress("unused")
    fun setFabColor(@ColorInt color: Int) {
        fab.backgroundTintList = ColorStateList.valueOf(color)
    }
}